itests/kit/node_unmanaged.go

```diff
@@ -93,271 +132,465 @@ func NewTestUnmanagedMiner(t *testing.T, full *TestFullNode, actorAddr address.A
 
+	ctx, cancel := context.WithCancel(ctx)
+
 	tm := TestUnmanagedMiner{
+		ctx:               ctx,
+		cancelFunc:        cancel,
 		t:                 t,
 		options:           options,
 		mockProofs:        mockProofs,
 		cacheDir:          cacheDir,
 		unsealedSectorDir: unsealedSectorDir,
 		sealedSectorDir:   sealedSectorDir,
 
-		unsealedSectorPaths: make(map[abi.SectorNumber]string),
-		cacheDirPaths:       make(map[abi.SectorNumber]string),
-		sealedSectorPaths:   make(map[abi.SectorNumber]string),
-		sealedCids:          make(map[abi.SectorNumber]cid.Cid),
-		unsealedCids:        make(map[abi.SectorNumber]cid.Cid),
-		sealTickets:         make(map[abi.SectorNumber]abi.SealRandomness),
+		committedSectors: make(map[abi.SectorNumber]sectorInfo),
 
 		ActorAddr:        actorAddr,
 		OwnerKey:         options.ownerKey,
 		FullNode:         full,
-		currentSectorNum: 101,
-		proofType:        make(map[abi.SectorNumber]abi.RegisteredSealProof),
+		currentSectorNum: abi.SectorNumber(actorIdNum * 100), // include the actor id to make them unique across miners and easier to identify in logs
 	}
 	tm.Libp2p.PeerID = peerId
 	tm.Libp2p.PrivKey = privkey
 
 	return &tm
 }
 
-func (tm *TestUnmanagedMiner) AssertNoPower(ctx context.Context) {
-	p := tm.CurrentPower(ctx)
-	tm.t.Logf("Miner %s RBP: %v, QaP: %v", tm.ActorAddr, p.MinerPower.QualityAdjPower.String(), p.MinerPower.RawBytePower.String())
-	require.True(tm.t, p.MinerPower.RawBytePower.IsZero())
+func (tm *TestUnmanagedMiner) Stop() {
+	tm.cancelFunc()
+	tm.AssertNoWindowPostError()
 }
 
-func (tm *TestUnmanagedMiner) CurrentPower(ctx context.Context) *api.MinerPower {
-	head, err := tm.FullNode.ChainHead(ctx)
-	require.NoError(tm.t, err)
+type OnboardOpt func(opts *onboardOpt) error
+type onboardOpt struct {
+	modifyNIActivationsBeforeSubmit func([]miner14.SectorNIActivationInfo) []miner14.SectorNIActivationInfo
+	requireActivationSuccess        bool
+}
 
-	p, err := tm.FullNode.StateMinerPower(ctx, tm.ActorAddr, head.Key())
-	require.NoError(tm.t, err)
+func WithModifyNIActivationsBeforeSubmit(f func([]miner14.SectorNIActivationInfo) []miner14.SectorNIActivationInfo) OnboardOpt {
+	return func(opts *onboardOpt) error {
+		opts.modifyNIActivationsBeforeSubmit = f
+		return nil
+	}
+}
 
-	return p
+func WithRequireActivationSuccess(requireActivationSuccess bool) OnboardOpt {
+	return func(opts *onboardOpt) error {
+		opts.requireActivationSuccess = requireActivationSuccess
+		return nil
+	}
 }
 
-func (tm *TestUnmanagedMiner) AssertPower(ctx context.Context, raw uint64, qa uint64) {
+// OnboardSectors onboards the specified number of sectors to the miner using the specified proof
+// type. If `withPieces` is true and the proof type supports it, the sectors will be onboarded with
+// pieces, otherwise they will be CC.
+//
+// This method is synchronous but each sector is prepared in a separate goroutine to speed up the
+// process with real proofs.
+func (tm *TestUnmanagedMiner) OnboardSectors(
```

Comment:
This pattern of onboard all here -> some are expected to fail and so wont be part of returned onboarded sectors here -> caller then asserts that the returned sectors match their expectation of "good sectors to onboard" did get a bit hard to reason about, especially as there are then code paths here where we check "if exit code == ok {} else {}" but don't fail the test if exit code is expected to be not okay.

I think it will be simpler if caller passes in a list of sector indices here that they expect to fail and we just assert on that here and account/assert on exit codes accordingly.

Reply:
it started off that way I think I changed it for 2 reasons: fewer arguments for common cases (less a problem now I have Options to pass in) and I wanted to check specific exit codes .. but in the end I think I removed that too!

I'll have a think about it, it felt dirty the first time I tried it but maybe now with options it'll be more pleasant

Reply:
Changed this all to just return `[]abi.SectorNumber` and it'll expect them all to succeed unless you specify a `WithExpectedExitCodes([]exitcode.ExitCode)`. It simplifies a few other things along the way.